pkg/agent/plugin/nodeattestor/devid/devid.go

```diff
@@ -0,0 +1,368 @@
+	spc "github.com/spiffe/spire/proto/spire/common"
+	"github.com/spiffe/spire/proto/spire/common/plugin"
+)
+
+const (
+	// defaultDevicePath is the value used when tpm_device_path is not set
+	defaultDevicePath = "/dev/tpmrm0"
+)
+
+func BuiltIn() catalog.Plugin {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.Plugin {
+	return catalog.MakePlugin(devid.PluginName, nodeattestor.PluginServer(p))
+}
+
+type ExternalConfig struct {
+	DevIDPrivPath string `hcl:"devid_priv_path"`
+	DevIDPubPath  string `hcl:"devid_pub_path"`
+	DevIDCertPath string `hcl:"devid_cert_path"`
+
+	AKPrivPath string `hcl:"ak_priv_path"`
+	AKPubPath  string `hcl:"ak_pub_path"`
+
+	DevicePath string `hcl:"tpm_device_path"`
+}
+
+type internalConfig struct {
+	trustDomain string
+	devicePath  string
+
+	devIDCert *x509.Certificate
+	devIDPub  []byte
+	devIDPriv []byte
+
+	checkDevIDResidency bool
+	akPub               []byte
+	akPriv              []byte
+}
+
+type Plugin struct {
+	nodeattestor.UnsafeNodeAttestorServer
+	log hclog.Logger
+
+	m sync.Mutex
+	c *internalConfig
+}
+
+func New() *Plugin {
+	return &Plugin{}
+}
+
+func (p *Plugin) FetchAttestationData(stream nodeattestor.NodeAttestor_FetchAttestationDataServer) error {
+	conf := p.getConfig()
+	if conf == nil {
+		return devid.Error("not configured")
+	}
+
+	// Open TPM connection and loads keys
+	tpm, err := loadTPMContext(conf, p.log)
+	if err != nil {
+		return fmt.Errorf("unable to load context: %w", err)
+	}
+	defer tpm.Close()
+
+	// Marshal attestation data
+	marshalledAttData, err := json.Marshal(devid.AttestationRequest{
+		DevIDCert: conf.devIDCert.Raw,
+		DevIDPub:  conf.devIDPub,
+
+		EKCert: tpm.EKCert,
+		EKPub:  tpm.EKPub,
+
+		AKPub: conf.akPub,
+
+		CertifiedDevID:         tpm.CertifiedDevID,
+		CertificationSignature: tpm.CertificationSignature,
+	})
+	if err != nil {
+		return fmt.Errorf("unable to marshall attestation data: %w", err)
+	}
+
+	// Send attestation request
+	err = stream.Send(&nodeattestor.FetchAttestationDataResponse{
+		AttestationData: &spc.AttestationData{
+			Type: devid.PluginName,
+			Data: marshalledAttData,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("unable to send attestation data: %w", err)
+	}
+
+	// Receive challenges
+	marshalledChallenges, err := stream.Recv()
+	if err != nil {
+		return fmt.Errorf("unable to receive challenge: %w", err)
+	}
+
+	challenges := &devid.ChallengeRequest{}
+	if err = json.Unmarshal(marshalledChallenges.Challenge, challenges); err != nil {
+		return fmt.Errorf("unable to unmarshall challenge: %w", err)
+	}
+
+	// Solve DevID challenge (verify the possession of the DevID private key)
+	devIDChallengeResp, err := tpm.SolveDevIDChallenge(challenges.DevID)
+	if err != nil {
+		return fmt.Errorf("unable to solve DevID challenge: %w", err)
+	}
+
+	// If DevID residency verification configured
+	var credActChallengeResp []byte
+	if conf.checkDevIDResidency && challenges.CredActivation != nil {
```

Comment:
> what happens in case checkDevIDResidency == true and CredActivation == nil

That state indicates that the agent was configured to perform the proof of residency but the server was configured to not require it. The server doesn't send any credential activation challenge to solve (`nil`).

> is it expected that it is a success challenge?

I wouldn't say that it is a successful resolution of the challenge. The challenge is not send by the sever, so the agent can (and must) ignore the "solving" function.